text/0000-min-rust-version.md

```diff
@@ -0,0 +1,179 @@
+became a warning instead of an error)
+
+`rust` will be a required field. For crates uploaded before introduction of this
+feature 2015 edition crates will imply `rust="1.0"` and 2018 ediiton will imply
+`rust = "1.30"`.
+
+It will be an error to use `rust="1.27"` and `edition="2018"`, but `rust="1.40"` and `edition="2015"` is a valid combination.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The describe functionality can be introduced in several stages:
+
+
+## First stage: dumb field
+
+At first the `rust` field can be simply a declarative optional field without any
+functionality behind it. The reason for it is to reduce implementation cost of
+the first stage to the minimum and ideally ship it as part of Rust 2018.
+It will also allow crate authors who care about MSRV to start mark their crates
+early.
+
+## Second stage: versions resolution
+
+`rust` field becomes required and cargo will add it as a constraint to dependency
+versions resolution. If user uses e.g. Rust 1.40 and uses crate `foo = "0.2"`, but
+all selected versions of `foo` specify MSRV e.g. equal 1.41 or bigger (or even
+nightly) `cargo` will issue an error.
+
+`rust` field value will be checked as well, on crate build `cargo` will check if
+all upstream dependencies can be built with the specified MSRV. (i.e. it will
+check if there is exists solution for given crates and Rust versions constraints)
+
+Yanked crates will be ignored in this process.
+
+Implementing this functionality hopefully will allow to close the debate regarding
+MSRV handling in crate versions and will allow crate authors to feel less
+restrictive about bumping their crate's MSRV. (though it can be a usefull
+convention for post-1.0 crates to bump minor version on MSRV change to allow
+publishing backports which fix serious issues using patch version)
+
+## Third stage: better crate checks
+
+Here we introduce two-level check for crates. First level will check if all used
+items were stabilised before or on given MSRV using `#[stable(since=version)]`
+attribute, issuing compile errors otherwise.
+
+Second level will try to build crate with the specified MSRV on `cargo publish`,
+i.e. words it will be required to install MSRV toolchain. (this check can be
+disabled using `--no-verify` option)
+
+While these two checks will not replace proper CI testing, they will help to
+reduce number of improper MSRV configuration to the minimum.
+
+Note that `rust` field must be equal to MSRV with default features for all
+supported targets.
+
+## Extension: nightly versions
+
+For some bleeding-edge crates which experience frequent breaks on Nightly updates
+(e.g. `rocket`) it can be useful to specify exact Nigthly version(s) on which
+crate can be built. One way to achieve this is by using the following syntax:
+- single version: rust = "nightly: 2018-01-01"
+- enumeration: "nightly: 2018-01-01, 2018-01-15"
+- (inclusive) range: "nightly: 2018-01-01..2018-01-15"
```

Comment:
Using `..` as inclusive range in Rust is gonna be confusing. Could we incorporate the [semver operators](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#inequality-requirements) instead?

```
nightly: >=2018-01-01, <=2018-01-15
```

(every new nightly is considered a major version for sure.)